embedded-io-async/src/impls/slice_mut.rs

```diff
@@ -20,7 +20,36 @@ impl Write for &mut [u8] {
             return Err(SliceWriteError::Full);
         }
         let (a, b) = mem::take(self).split_at_mut(amt);
-        a.copy_from_slice(&buf[..amt]);
+        a.copy_from_slice(buf.split_at(amt).0);
+        *self = b;
+        Ok(amt)
+    }
+}
+
+/// Write is implemented for `&mut [MaybeUninit<u8>]` by copying into the slice, initializing
+/// & overwriting its data.
+///
+/// Note that writing updates the slice to point to the yet unwritten part.
+/// The slice will be empty when it has been completely overwritten.
+///
+/// If the number of bytes to be written exceeds the size of the slice, write operations will
+/// return short writes: ultimately, `Ok(0)`; in this situation, `write_all` returns an error of
+/// kind `ErrorKind::WriteZero`.
+impl Write for &mut [MaybeUninit<u8>] {
+    #[inline]
+    async fn write(&mut self, buf: &[u8]) -> Result<usize, Self::Error> {
+        let amt = core::cmp::min(buf.len(), self.len());
+        if !buf.is_empty() && amt == 0 {
+            return Err(SliceWriteError::Full);
+        }
+        let (a, b) = mem::take(self).split_at_mut(amt);
+        buf.split_at(amt)
```

Comment:
maybe core::ptr::copy_nonoverlapping has a better chance at getting optimized than iterator loops. Or maybe copy_from_slice but i'm not sure if it's possible to use soundly.

Reply:
also `buf.split_at(amt).0` can be written more readably as `buf[..amt]`

Reply:
> but that requires unsafe code

that's not a problem

> they behave differently though - .split_at() splits the lifetime and also works in const contexts, so i think it is preferred over slicing with []

the lifetime is not an issue with readonly `&[u8]` because aliasing is allowed, and slicing is also allowed in const.

Reply:
from what i can tell, slicing is not allowed in const contexts.

this code snippet compiles just fine:
```
const HELLO_WORLD_BYTES: &[u8] = b"hello world";
const HELLO_BYTES_SPLIT_AT: &[u8] = &HELLO_WORLD_BYTES.split_at(4).0;
fn main() { panic!() }
```

whereas this code snippet fails to compile:
```
const HELLO_WORLD_BYTES: &[u8] = b"hello world";
const HELLO_BYTES_SLICE_OPERATOR: &[u8] = &HELLO_WORLD_BYTES[..4];
fn main() { panic!() }
```

compiler error:
```
error[E0015]: cannot call non-const operator in constants
calls in constants are limited to constant functions, tuple structs and tuple variants
```